File: ruby/test/support/fake_catalog.rb
class FakeCatalog < SupermarketCatalog

  def initialize
    @products = {}
    @prices = {}
  end

  def add_product(product, price)
    @products[product.name] = product
    @prices[product.name] = price
  end

  def unit_price(p)
    @prices.fetch(p.name)
  end

end

File: elixir/test/support/fake_catalog.ex
defmodule FakeCatalog do
  defstruct [:products, :prices]

  def new, do: %__MODULE__{products: %{}, prices: %{}}

  defimpl Supermarket.Model.SupermarketCatalog do
    def add_product(catalog, product, price) do
      catalog
      |> Map.update!(:products, &Map.put(&1, product.name, product))
      |> Map.update!(:prices, &Map.put(&1, product.name, price))
    end

    def get_unit_price(catalog, product) do
      catalog.prices[product.name]
    end
  end
end
